<!--Copyright (c) 2018 Great Ormond Street Hospital for Children NHS Foundation
Trust & Birmingham Women's and Children's NHS Foundation Trust

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-->

<html>
<head>
<title>Technical Report</title>
<!--This is the meta information which is good for SEO-->
<!-- Bootstrap -->
<!--Add keywords that describe your project between the quotation marks. This helps search engines find and index your content-->


<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

</head>


<body>

<page style="margin: 0 auto; width:21cm;"size="A4">

<div class="container">
    <div class="row">
        <div class="col-sm-6" style="padding: 0">
            <strong>
            <p style="margin-bottom: 0">West Midlands Regional Genetics Laboratory</p>
            <p style="margin-bottom: 0">Birmingham Women's and Children's Hospital</p>
            <p style="margin-bottom: 0">NHS Foundation Trust</p>
            <p style="margin-bottom: 0">Mindelsohn Way</p>
            <p style="margin-bottom: 10">B15 2QQ</p>
            </strong>
        </div>
        <div class="col-sm-6">
        </div>
    </div>
</div>

<hr style="margin: 0px; height: 0px; border: 2px solid black; color: black; background: black;">
<p style="margin: auto; text-align: center;"><strong>TECHNICAL REPORT</strong></p>
<hr style="margin: 0px; height: 0px; border: 2px solid black; color: black; background: black;">

<p style="font-size: 9pt">Our reference: {{report.ir_family.participant_family.proband.lab_number}}</p>

<div class="container">
    <div class="row">
        <div class="col-sm-6">
            <p style="font-size: 9pt; margin-bottom: 0">Referring Clinician:</p>
            <p style="font-size: 12pt">{{report.ir_family.participant_family.clinician.name}}</p>
        </div>
        <div class="col-sm-6">

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg .tg-0ord{text-align:right}
.tg .tg-lqy6{text-align:right;vertical-align:top}
.tg .tg-yw4l{vertical-align:top}
.tgd{padding-left:5px}
</style>
<table class="tg">
  <tr>
    <td class="tgd tg-0ord">Patient Name:</td>
    <td class="tgd tg-031e">{{report.ir_family.participant_family.proband.forename}} {{report.ir_family.participant_family.proband.surname}}</td>
  </tr>
  <tr>
    <td class="tgd tg-0ord">Date of Birth:</td>
    <td class="tgd tg-031e">{{report.ir_family.participant_family.proband.date_of_birth|date:"d/m/Y"}}</td>
  </tr>
  <tr>
    <td class="tgd tg-lqy6">Sex:</td>
    <td class="tg-yw4l">{{report.ir_family.participant_family.proband.gender}}</td>
  </tr>
  <tr>
    <td class="tg-lqy6">NHS No.:</td>
    <td class="tg-yw4l">{{report.ir_family.participant_family.proband.nhs_number}}</td>
  </tr>
  <tr>
    <td class="tg-lqy6">Hospital No.:</td>
    <td class="tg-yw4l">{{report.ir_family.participant_family.proband.cgu_number}}</td>
  </tr>
  <tr>
    <td class="tg-lqy6">100,000 Genomes ID:</td>
    <td class="tg-yw4l">{{report.ir_family.participant_family.proband.gel_id}}</td>
  </tr>
</table>

        </div>
    </div>
</div>

<hr style="margin: 0px; height: 0px; border: 2px solid black; color: black; background: black;">

    <h3>Recruiting Disease</h3>
    <p>
    {{ report.ir_family.participant_family.proband.recruiting_disease }}
    </p>





    <h3>Variant Details</h3>

    {% if outcome != "positive" %}
        <strong>No variants have been detected.</strong>
    {% elif outcome == "positive" %}
        <small>
            <table class="table table-bordered table-condensed" style="font-size:10px;">
                <thead>
                    <tr>
                        <th>Gene</th>
                        <th>Zygosity</th>
                        <th>Inheritance</th>
                        <th>HGVS description</th>
                        <th>Position ({{build}})</th>
                        <th>Classification</th>
                        <th>VEP Consequence</th>
                        </th>
                </thead>
                <tbody>
                    {% for variant in reported_variants %}
                        <tr>
                            <td>{{variant.proband_variant.get_transcript.gene}}</td>
                            <td>{{variant.proband_variant.zygosity}}</td>
                            <td>{{variant.proband_variant.inheritance}}</td>
                            <td>
                                {{variant.proband_variant.get_transcript_variant.hgvs_c}}
                                {% if variant.proband_variant.get_transcript_variant.hgvs_p %}
                                    <br>
                                    {{variant.proband_variant.get_transcript_variant.hgvs_p}}
                                {% endif %}
                            </td>
                            <td>{{variant.proband_variant.get_transcript_variant.variant.position}}</td>
                            <td>{{variant.classification}}</td>
                            <td>{{variant.proband_variant.get_ptv.effect}}</td>
                        </tr>
        {% endfor %}
        </tbody>
    </table>
        </small>
    {% endif %}



    <h3>Gene Panels and Read Depth Coverage</h3>

    <table class="table table-condensed">
        <thead>
            <tr>
                <th>Panel</th>
                <th style="text-align:right">No. Genes</th>
                <th style="text-align:right">Avg. Panel Coverage</th>
                <th style="text-align:right">% >15x</th>
            </th>
        </thead>
        <tbody>

        {% for panel, num_genes in panels.items %}
            <tr>
            <td>{{panel.panel}}</td>
            <td style="text-align: right">{{num_genes}}</td>
            <td style="text-align: right">{{ panel.average_coverage }}x</td>
            <td style="text-align: right">{% widthratio panel.proportion_above_15x 1 100 %}%</td>
        </tr>
    {% endfor %}
        </tbody>
    </table>


    <h4>Genes with low coverage*: </h4>
        {% for panel, num_genes in panels.items %}
        {% if forloop.first %}

        <p style="text-align:justify">
        <small>
        {{ panel.genes_failing_coverage }}
    </small>
        </p>

        {% endif %}
        {% endfor %}
    <h6>* Low coverage = less than 95% of the gene has a minimum read depth coverage of 15x.</h6>

</page>

</body>

</html>
