<!--Copyright (c) 2018 Great Ormond Street Hospital for Children NHS Foundation
Trust & Birmingham Women's and Children's NHS Foundation Trust

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
-->
{% load static %}
{% load bootstrap3 %}

<html>
<head>
<title>Technical Report</title>
<!--This is the meta information which is good for SEO-->
<!-- Bootstrap -->
<!--Add keywords that describe your project between the quotation marks. This helps search engines find and index your content-->
<link rel="shortcut icon" type="image/png" href="{% static 'gel2mdt/favicon/favicon-96x96.png' %}"/>
{% block head-css %}
    <script src="{% static 'gel2mdt/js/jquery-3.1.1.js' %}"></script>
<script src="{% static 'gel2mdt/js/bootstrap.js' %}"></script>
<link href="{% static 'gel2mdt/css/bootstrap.css' %}" rel="stylesheet">
<link href="{% static 'gel2mdt/css/heroic-features.css' %}" rel="stylesheet">
<link href="{% static 'gel2mdt/DataTables/datatables.min.css' %}" rel="stylesheet">
<link href="{% static 'gel2mdt/css/jumbotron.css' %}" rel="stylesheet">
<link href="{% static 'gel2mdt/css/bootstrap-responsive.css' %}" rel="stylesheet">
<script type="text/javascript" src="{% static 'gel2mdt/DataTables/datatables.min.js' %}"></script>
<script type="text/javascript" src="{% static 'gel2mdt/js/all.js' %}"></script>

<link rel="stylesheet" type="text/css" href="{% static 'gel2mdt/css/report.css' %}">

    {% endblock %}
{% block javascript %}
{% endblock %}


</head>


<body>

<page size="A4">


    <h2>Rare Disease Whole Genome Sequencing - Technical Report</h2>


    <h3>Patient name: <strong>{{ report.ir_family.participant_family.proband.forename}} {{report.ir_family.participant_family.proband.surname}}</strong></h3>

    <h3><strong>Sample Details</strong></h3>
    <table class="table" style="table-layout:fixed">
        <col style="width:25%", span="4" />
        <tr>
            <th>Laboratory No.:</th>
            <th> {{ report.ir_family.participant_family.proband.lab_number }}</th>
             <th>GeL ID:</th>
            <th> {{ report.ir_family.participant_family.proband.gel_id }}</th>
       <tr>
             <th>Date of Birth:</th>
            <th> {{ report.ir_family.participant_family.proband.date_of_birth|date:"d/m/Y" }}</th>
             <th>NHS No.:</th>
            <th> {{ report.ir_family.participant_family.proband.nhs_number }}</th>
      </tr>
    </table>



    <h3>Recruiting Disease</h3>
    <p>
    {{ report.ir_family.participant_family.proband.recruiting_disease }}
    </p>

    <hr>

    <h3>Test Methodology</h3>
    <p style="text-align:justify">
    Whole genome sequencing by the 100,000 Genomes Project with analysis of
    selected gene panels (see below). Please note that the sensitivity of this test
    is limited by the types of detectable pathogenic variants, regions of low read
    depth coverage and incomplete ascertainment of disease-gene associations. More
    detailed information regarding read depth coverage is available on request.
    Please contact the laboratory and quote the Laboratory No. for more information.
    {% if outcome == "positive" %}
        Variants are classified using the ACMG/AMP guidelines
        (Richards et al 2015 Genet Med) and confirmed using in-house Sanger sequencing.
    {% endif %}
    </p>


    <h3>Variant Details</h3>

    {% if outcome != "positive" %}
        <strong>No variants have been detected.</strong>
    {% elif outcome == "positive" %}
        <small>
            <table class="table table-bordered table-condensed" style="font-size:10px;">
                <thead>
                    <tr>
                        <th>Gene</th>
                        <th>Zygosity</th>
                        <th>Inheritance</th>
                        <th>HGVS description</th>
                        <th>Position ({{build}})</th>
                        <th>Classification</th>
                        <th>VEP Consequence</th>
                        </th>
                </thead>
                <tbody>
                    {% for variant in reported_variants %}
                        <tr>
                            <td>{{variant.proband_variant.get_transcript.gene}}</td>
                            <td>{{variant.proband_variant.zygosity}}</td>
                            <td>{{variant.proband_variant.inheritance}}</td>
                            <td>
                                {{variant.proband_variant.get_transcript_variant.hgvs_c}}
                                {% if variant.proband_variant.get_transcript_variant.hgvs_p %}
                                    <br>
                                    {{variant.proband_variant.get_transcript_variant.hgvs_p}}
                                {% endif %}
                            </td>
                            <td>{{variant.proband_variant.get_transcript_variant.variant.position}}</td>
                            <td>{{variant.classification}}</td>
                            <td>{{variant.proband_variant.get_ptv.effect}}</td>
                        </tr>
        {% endfor %}
        </tbody>
    </table>
        </small>
    {% endif %}



    <h3>Gene Panels and Read Depth Coverage</h3>

    <table class="table table-condensed">
        <thead>
            <tr>
                <th>Panel</th>
                <th style="text-align:right">No. Genes</th>
                <th style="text-align:right">Avg. Panel Coverage</th>
                <th style="text-align:right">% >15x</th>
            </th>
        </thead>
        <tbody>

        {% for panel, num_genes in panels.items %}
            <tr>
            <td>{{panel.panel}}</td>
            <td style="text-align: right">{{num_genes}}</td>
            <td style="text-align: right">{{ panel.average_coverage }}x</td>
            <td style="text-align: right">{% widthratio panel.proportion_above_15x 1 100 %}%</td>
        </tr>
    {% endfor %}
        </tbody>
    </table>


    <h4>Genes with low coverage*: </h4>
        {% for panel, num_genes in panels.items %}
        {% if forloop.first %}

        <p style="text-align:justify">
        <small>
        {{ panel.genes_failing_coverage }}
    </small>
        </p>

        {% endif %}
        {% endfor %}
    <h6>* Low coverage = less than 95% of the gene has a minimum read depth coverage of 15x.</h6>

</page>

</body>

</html>
